lib/pact/provider/generators/provider_state.rb

```diff
@@ -0,0 +1,58 @@
+require 'pact/provider/generators'
+
+module Pact
+  module Provider
+    class ProviderStateGenerator
+
+      
+      # rewrite of https://github.com/DiUS/pact-jvm/blob/master/core/support/src/main/kotlin/au/com/dius/pact/core/support/expressions/ExpressionParser.kt#L27
+      VALUES_SEPARATOR = ","
+      START_EXPRESSION = "\${"
+      END_EXPRESSION = '}'
+      def parse_expression expression, params 
+        
+        return_string = []
+
+        buffer = expression;
+        # initial value
+        position = buffer.index(START_EXPRESSION)
+        
+        while (position && position >= 0) 
+          if (position > 0) 
+            # add string 
+            return_string.push(buffer[0...position])
+          end
+          end_position = buffer.index(END_EXPRESSION, position)
+          if (end_position < 0)
+            raise "Missing closing brace in expression string \"#{$value}\""
+          end
+
+          variable = ""
+
+          if (end_position - position > 2)
+            expression = params[buffer[position+2...end_position]] || ""
```

Comment:
Would it be better to loudly error/explode if the key is not found in the params?

Say the provider state endpoint returns a map with "auth_token" and your contract has a generator looking for "customer_auth_token", debugging this is frustrating.  Better to notify the user with some useful feedback?  Rather than have the contract fail for seemingly unrelated reasons.